src/_Derivatives/subexpressions.jl

```diff
@@ -14,78 +14,130 @@ function list_subexpressions(nd::Vector{NodeData})
     return sort(collect(indices))
 end
 
-# Order the subexpressions which main_expressions depend on such that we can
-# run forward mode in this order.
-function order_subexpressions(
-    main_expressions::Vector{Vector{NodeData}},
+"""
+Return the list of dependent subexpressions. Similar to `list_subexpressions`,
+but without the unnecessary `Set{Int}` -> `Vector{Int}` -> `sort`.
+"""
+function _subexpressions(nd::Vector{NodeData})
+    return Int[n.index for n in nd if n.nodetype == SUBEXPRESSION]
+end
+
+"""
+Return `G[node]`, lazily building it from `subexpressions` if it is not yet
+assigned.
+"""
+function _children(
+    G::Vector{Vector{Int}},
     subexpressions::Vector{Vector{NodeData}},
+    node::Int,
 )
-    num_sub = length(subexpressions)
-    computed = falses(num_sub)
-    dependencies = Dict{Int,Vector{Int}}()
-    to_visit = collect(num_sub+1:num_sub+length(main_expressions))
-    # For each subexpression k, the indices of the main expressions that depend
-    # on k, possibly transitively.
-    depended_on_by = [Set{Int}() for i in 1:num_sub]
-
-    while !isempty(to_visit)
-        idx = pop!(to_visit)
-        if idx > num_sub
-            subexpr = list_subexpressions(main_expressions[idx-num_sub])
-        else
-            computed[idx] && continue
-            subexpr = list_subexpressions(subexpressions[idx])
-            computed[idx] = true
-        end
-        dependencies[idx] = subexpr
-        for k in subexpr
-            if idx > num_sub
-                push!(depended_on_by[k], idx - num_sub)
-            end
-            push!(to_visit, k)
-        end
+    if !isassigned(G, node)
+        G[node] = _subexpressions(subexpressions[node])
     end
+    return G[node]
+end
 
-    # Now order dependencies.
-    I = Int[]
-    J = Int[]
-    for (idx, subexpr) in dependencies
-        for k in subexpr
-            push!(I, idx)
-            push!(J, k)
+"""
+    _topological_sort(starts, G::Vector{Vector{Int}})
+
+Return a topologically sorted list of nodes in `G` that need to be evaluated
```

Comment:
Explain the structure of G. G[i] is a list of "out" edges from node i, possibly with duplicate edges, some of which might be undef and computed lazily?